heartbeat/monitors/active/dialchain/tls.go

```diff
@@ -35,24 +36,36 @@ import (
 //        "rtt": { "handshake": { "us": ... }}
 //    }
 //  }
-func TLSLayer(cfg *transport.TLSConfig, to time.Duration) Layer {
+func TLSLayer(cfg *transport.TLSConfig, to time.Duration, meta *transport.DialerMeta) Layer {
 	return func(event common.MapStr, next transport.Dialer) (transport.Dialer, error) {
 		var timer timer
 
 		// Wrap next dialer so to start the timer when 'next' returns.
 		// This gets us the timestamp for when the TLS layer will start the handshake.
 		next = startTimerAfterDial(&timer, next)
 
-		dialer, err := transport.TLSDialer(next, cfg, to)
+		dialer, err := transport.TLSMetaDialer(next, cfg, to, meta)
 		if err != nil {
 			return nil, err
 		}
 
 		return afterDial(dialer, func(conn net.Conn) (net.Conn, error) {
 			// TODO: extract TLS connection parameters from connection object.
-
 			timer.stop()
 			event.Put("tls.rtt.handshake", look.RTT(timer.duration()))
+
+			connState := meta.TLS.ConnectionState()
```

Comment:
We really need all the wiring up? There is quite some indirection here, I'd rather keep out if possible. The `conn` connection object should be a tls connection (as is being produced by TLSDialer). We should read the connection state from the `conn` parameter if possible.

Reply:
Hmmm, what exactly are you thinking here? `tls.Conn` in [crypto/tls/conn.go](https://golang.org/src/crypto/tls/conn.go) is not a super-type (is that the go lingo?) of `net.Conn`, so cannot be cast. It simply nests a `net.Conn` as a property.

AFAICT this means we have to use this sort of indirection. Would love a simpler alternative though. Do I misunderstand?

Reply:
tls.Conn wraps another connection, because TLS sits on top of another network layer. But tls.Conn implements `net.Conn` itself. In go we do not have inheritance or subtype relationships. But `tls.Conn` structurally satisfies the `net.Conn` interface. When building layers of dialers, we have to fix the return values of dialers, due to lack of generics and sub-typing (no co-/contra-variance on methods). The common denominator for dialers is given by:

```
type Dialer interface {
	Dial(network, address string) (net.Conn, error)
}
```

As `net.Conn` is an interface, each layer will wrap it's input layer + use a concrete type which is finally promoted to the interface value being returned.

This is reflected in the DialerChain type:

```
type DialerChain struct {
	Net    NetDialer
	Layers []Layer
}
```

The NetDialer produces the most low-level connection object. Each layer will wrap the current connection object, returning a new connection object.

As we want to do some measurements before and after a network layer becomes active, I introduced the `beforeDial` and `afterDial` wrappers, so to reduce common boilerplate and amount of wrapper functions needed. Again, lack of generics and sub-type relationships we do enforce the use of `net.Conn`. But in general, once a dialer-chain is setup, we have this layout:

```
+----+-----------------------+----+ net.Conn
     |  After Layer II       |
     +-----------------------+ Layer II Conn type
     |  Dial Layer II        |
     +-----------------------+
     |  Before Layer II      |
+---------------------------------+  net.Conn
     |  After Layer I        |
     +-----------------------+ Layer I Conn type
     |  Dial Layer I         |
     +-----------------------+
     |  Before Layer I       |
+---------------------------------+ net.Conn
     |  Net Layer (e.g. TCP) |
     +-----------------------+
```

A Layer must make no assumptions about the underlying layer. A layer must assume the connection object to wrap to be of type `net.Conn`. But within a layer (dial + after steps), we know the concrete type returned by the dialer and to be consumed by the 'after'-handler. Due to limitations of the type-system, we are forced to cast the interface value to the concrete type, though.

In case of the TLS layer, we use transport.TLSDialer. The concrete type returned  by `(*TLSDialer).Dial` is `*tls.Conn`. => we can cast the interface value to `*tls.Conn` in the 'after'-handler. As we know the concrete type in the 'after'-handler, we don't need another behind-the-scenes type capturing the concrete type. Admittedly, it's more type safe, but it's somewhat more involved and somewhat over-engineered for something that should be simple (as the after-handler is the only consumer that requires to access the tls state).

Reply:
Thanks for the very detailed writeup. I'm realizing that my confusion was in typing `conn.(tls.Conn)` instead of `conn.(*tls.Conn)`. That was the root of my assumption that the interface wasn't implemented.

Thanks for the detailed diagram of the dialer stack, that's very helpful!

Reply:
We should move that diagram to the code docs somewhere.